### PR TITLE
Istep 13 9

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

**Comment by macpijan** [comment]
@miczyg1 please review

File: src/include/cpu/power/istep_13.h [hunk 1]
```diff
@@ -95,17 +100,20 @@ extern mcbist_data_t mem_data;
...
-static inline uint64_t ps_to_nck(uint64_t ps)
+static inline uint64_t tck_in_ps(void)
 {
 	/*
 	 * Speed is in MT/s, we need to divide it by 2 to get MHz.
 	 * tCK(avg) should be rounded down to the next valid speed bin, which
 	 * corresponds to value obtained by using standardized MT/s values.
 	 */
-	uint64_t tck_in_ps = 1000000 / (mem_data.speed / 2);
+	return 1000000 / (mem_data.speed / 2);
```
Comment (miczyg1) [inline src/include/cpu/power/istep_13.h:110]:
```suggestion
	return PSEC_PER_USEC / (mem_data.speed / 2);
```

We define a value but do not use it? Or is it something different?

File: src/include/cpu/power/istep_13.h [hunk 2]
```diff
@@ -95,17 +100,20 @@ extern mcbist_data_t mem_data;
...
+}
 
+static inline uint64_t ps_to_nck(uint64_t ps)
+{
 	/* Algorithm taken from JEDEC Standard No. 21-C */
-	return ((ps * 1000 / tck_in_ps) + 974) / 1000;
+	return ((ps * 1000 / tck_in_ps()) + 974) / 1000;
```
Comment (miczyg1) [inline src/include/cpu/power/istep_13.h:116]:
```suggestion
	return ((ps * PSEC_PER_NSEC / tck_in_ps()) + 974) / PSEC_PER_NSEC;
```

Applicable or not?

File: src/include/cpu/power/istep_13.h [hunk 3]
```diff
@@ -116,11 +124,64 @@ static inline uint64_t mtb_ftb_to_nck(uint64_t mtb, int8_t ftb)
...
+{
+	/* Indirect registers have different stride than the direct ones. */
+	unsigned mul = (scom & PPC_BIT(0)) ? 0x400 : 0x40;
+	scom_and_or_for_chiplet(mcs, scom + mca * mul, and, or);
+}
+
+static inline void dp_mca_and_or(chiplet_id_t mcs, int dp, int mca,
+                                 uint64_t scom, uint64_t and, uint64_t or)
+{
+	mca_and_or(mcs, mca, scom + dp * 0x40000000000, and, or);
```
Comment (miczyg1) [inline src/include/cpu/power/istep_13.h:167]: `dp` could be bit-shifted? Multiplying by a big hex value like this is weird. What is the exact meaning of this multiplication?

File: src/soc/ibm/power9/istep_13_9.c [hunk 4]
```diff
@@ -0,0 +1,751 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <cpu/power/istep_13.h>
+#include <console/console.h>
+#include <timer.h>
+
+static int test_dll_calib_done(chiplet_id_t id, int mca_i, bool *do_workaround)
+{
+	uint64_t status = mca_read(id, mca_i, 0x8000C0000701103F);
```
Comment (miczyg1) [inline src/soc/ibm/power9/istep_13_9.c:9]: Please use the preprocessor to define register names, using hardcoded values is ugly.

File: src/soc/ibm/power9/istep_13_9.c [hunk 5]
```diff
@@ -0,0 +1,751 @@
...
+			[51]  ADR_DLL_CAL_GOOD ==       1
+			[52]  ADR_DLL_CAL_ERROR ==      0
+			[53]  ADR_DLL_CAL_ERROR_FINE == 0) break    // success
+	if (IOM0.DDRPHY_PC_DLL_ZCAL_CAL_STATUS_P0
+			[49]  DP_DLL_CAL_ERROR ==       1 |
+			[50]  DP_DLL_CAL_ERROR_FINE ==  1 |
+			[52]  ADR_DLL_CAL_ERROR ==      1 |
+			[53]  ADR_DLL_CAL_ERROR_FINE == 1) break and do the workaround
+	*/
+	if ((status & PPC_BITMASK(48, 53)) == (PPC_BIT(48) | PPC_BIT(51))) {
```
Comment (miczyg1) [inline src/soc/ibm/power9/istep_13_9.c:24]: You may pass the status to a simple function that checks for good bits. These bits can have their names too in preprocessor or something.

File: src/soc/ibm/power9/istep_13_9.c [hunk 6]
```diff
@@ -0,0 +1,751 @@
...
+		/* DLL calibration finished without errors */
+		return 1;
+	}
+
+	if (status & (PPC_BIT(49) | PPC_BIT(50) | PPC_BIT(52) | PPC_BIT(53))) {
```
Comment (miczyg1) [inline src/soc/ibm/power9/istep_13_9.c:29]: Same here, the function could check for error bits

File: src/soc/ibm/power9/istep_13_9.c [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+		IOM0.IOM_PHY0_DDRPHY_FIR_MASK_REG
+			  [54]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_0 = 0   // recoverable_error (0,1,0)
+			  [55]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_1 = 0   // recoverable_error (0,1,0)
+			  [57]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_3 = 0   // recoverable_error (0,1,0)
+			  [58]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_4 = 0   // recoverable_error (0,1,0)
+			  [59]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_5 = 0   // recoverable_error (0,1,0)
+			  [60]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_6 = 0   // recoverable_error (0,1,0)
+			  [61]  IOM_PHY0_DDRPHY_FIR_REG_DDR_FIR_ERROR_7 = 0   // recoverable_error (0,1,0)
+		*/
+		/* FIXME: how is PHY numbering related to MCA numbering? mca_i/2? */
```
Comment (miczyg1) [inline src/soc/ibm/power9/istep_13_9.c:262]: Maybe a good question to mailing list?

File: src/soc/ibm/power9/istep_13_9.c [hunk 8]
```diff
@@ -0,0 +1,751 @@
...
+		//mca_and_or(id, mca_i, 0x07011003,
+		mca_and_or(id, 0, 0x07011003,
+		           ~(PPC_BITMASK(54, 55) | PPC_BITMASK(57, 61)),
+		           0);
+	}
+}
+
+/*
+ * Can't protect with do..while, this macro is supposed to exit 'for' loop in
+ * which it is invoked. As a side effect, it is used without semicolon.
```
Comment (miczyg1) [inline src/soc/ibm/power9/istep_13_9.c:280]: "If only not the IF" and the braces, it could end with semicolon

**Review by miczyg1 — CHANGES_REQUESTED** [review]

**Comment by macpijan** [comment]
Already integrated in: https://github.com/3mdeb/coreboot/pull/49

**Pull request closed without merging**